tools/scripts/ffi.toml

```diff
@@ -312,4 +313,137 @@ else
+if is_path_exists fuchsia
+    echo "NOTE: fuchsia already exists"
+else
+    exec --fail-on-error git clone --depth 1 https://fuchsia.googlesource.com/fuchsia
+    exec ln -s Cargo.toml.crates-io fuchsia/tools/cargo-gnaw/Cargo.toml
+end
+
+# Install GNaw locally in the bin directory (alongside GN)
+# Since the parent directory is vendored, we have no choice but to vendor here, too
+mkdir .cargo
+writefile .cargo/config "source = { crates-io.replace-with = 'vendored-sources', vendored-sources.directory = 'vendor' }"
+exec cargo vendor --manifest-path fuchsia/tools/cargo-gnaw/Cargo.toml
+exec cargo install --path fuchsia/tools/cargo-gnaw --root .
+
+# Ensure everything works
+exec --fail-on-error ./bin/gn --version
+exec --fail-on-error ./bin/gnaw --help
+
+echo "*** GN Tools Successfully Installed! Ignore all previous messages. ***"
+'''
+
+[tasks.gn-gen]
+description = "Generate GN build rules for ICU4X"
+category = "ICU4X Development"
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+
+third_party_tools = is_path_exists ffi/gn/third_party_tools
+assert ${third_party_tools} "The GN third-party tools are not installed.\n*** Please run 'cargo make gn-install' ***"
+
+cd ffi/gn
+exec rm -f icu4x_root
+exec ln -s ../.. icu4x_root
+echo "NOTE: Created a temporary symbolic link at /ffi/gn/icu4x_root"
+
+pwd = pwd
+rm -rf vendor
+output = exec --fail-on-error cargo vendor
+writefile .cargo/config "# This file is part of ICU4X. For terms of use, please see the file\n"
+appendfile .cargo/config "# called LICENSE at the top level of the ICU4X source tree\n"
+appendfile .cargo/config "# (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).\n\n"
+appendfile .cargo/config "# AUTO-GENERATED in ffi.toml\n"
+appendfile .cargo/config ${output.stdout}
+exec --fail-on-error ./third_party_tools/bin/gnaw --manifest-path "${pwd}/Cargo.toml" --project-root ${pwd} --output ${pwd}/icu4x/BUILD.gn --skip-root --gn-bin ${pwd}/third_party_tools/bin/gn
+
+exec rm -f icu4x_root
+echo "NOTE: Removed the temporary symbolic link /ffi/gn/icu4x_root"
+'''
+
+[tasks.verify-gn-gen]
+description = "Verify that checked-in GN rules are up to date"
+category = "ICU4X Development"
+dependencies = [
+    "gn-gen",
+]
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+code = exec --get-exit-code git diff --exit-code -- ffi/gn
+if ${code}
+    trigger_error "GN rules need to be regenerated. Please run `cargo make gn-install`, then `cargo make gn-gen`, and commit."
+end
+'''
+
+[tasks.gn-build]
+description = "Build the GN version of ICU4X"
+category = "ICU4X Development"
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+
+third_party_tools = is_path_exists ffi/gn/third_party_tools
+assert ${third_party_tools} "The GN third-party tools are not installed.\n*** Please run 'cargo make gn-install' ***"
+
+cd ffi/gn
+exec rm -f icu4x_root
+exec ln -s ../.. icu4x_root
+echo "NOTE: Created temporary symbolic link at /ffi/gn/icu4x_root"
```

Comment:
nit: comment about why this is needed

Reply:
i think i can guess why

i kinda wish gn/gnaw (whoever's fault this is) could be given more paths so this isn't necessary, but whatever

if we can avoid this we should try

Reply:
GN wants a root directory, and you can't get out of the sandbox by going up (`..`) more directory levels. We could make the repository root be the GN root, but then I'm pretty sure we'd need to put the GN config files at the repository root. But I agree; I would really like to get rid of the symlink.

Maybe @zanderso has suggestions here?

Reply:
yeah i was hoping there's a way to either:

 - run gn from the root but tell it to write to a folder
 - run gn from the folder but tell it where the root is

directory-agnosticness is typically good to have in a tool

(not saying we should fix this, but hoping it's already fixed and if not they'll take a bug report)

Reply:
Not 100% sure, but it sounds like the `--root` switch to `gn` might be helpful. `gn help --root` will tell you about it.

Reply:
Cool; if I put the `.gn` file at the top level, then the build paths are relative to the top level, but I can add configurations inside of `.gn` to keep the BUILD.gn files all inside //fii/gn.

Reply:
sweet that seems fine